src/coreclr/tools/ILVerification/ILImporter.Verify.cs

```diff
@@ -2122,18 +2122,18 @@ void ImportStoreField(int token, bool isStatic)
             ClearPendingPrefix(Prefix.Volatile);
 
             var value = Pop();
-
             var field = ResolveFieldToken(token);
-
             TypeDesc instance;
+
             if (isStatic)
             {
                 Check(field.IsStatic, VerifierError.ExpectedStaticField);
 
                 instance = null;
 
-                if (field.IsInitOnly)
-                    Check(_method.IsStaticConstructor && field.OwningType == _method.OwningType, VerifierError.InitOnly);
+                Check(!field.IsInitOnly ||
+                      (field.OwningType == _method.OwningType && _method.IsStaticConstructor) ||
+                      (_method.IsStaticConstructor && field.OwningType == _method.OwningType), VerifierError.InitOnly);
```

Comment:
This just added duplicate code and made the whole thing much harder to read. I think the original version was just fine.